### ISSUE TITLE
Exporting PyTorch code for use in RASR

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1289]
For reference, see #1120. This is probably only relevant for the RF, because otherwise there is nothing RETURNN-specific about it (no RETURNN code influences this when you don't use RETURNN code but pure PT instead).

Do we need to support TorchScript graph capture (`torch.jit.script`/`torch.jit.trace`) for ONNX?

Tracing (`torch.jit.trace`) should anyway always work, except it would not cover dynamic control flow.

`torch.compile` for ONNX will be the future but does not work yet?

Maybe also [OpenXLA](https://opensource.googleblog.com/2023/03/openxla-is-ready-to-accelerate-and-simplify-ml-development.html) instead of ONNX.

Any other reasonable way to get PT models running within RASR? We can also simply use the RASR-Python bridge.

Another RF backend just specifically for TorchScript export?

(Some earlier discussion: https://github.com/rwth-i6/returnn/issues/1120#issuecomment-1484230901, https://github.com/rwth-i6/returnn/issues/1120#issuecomment-1484253666, https://github.com/rwth-i6/returnn/issues/1120#issuecomment-1484714695)

**Comment by JackTemaki** [comment]
> For reference, see #1120. This is probably only relevant for the RF, because otherwise there is nothing RETURNN-specific about it (no RETURNN code influences this when you don't use RETURNN code but pure PT instead).
> 
> Do we need to support TorchScript graph capture (`torch.jit.script`/`torch.jit.trace`) for ONNX?

I think this highly depends on the timeline. In order to work on RESCALE I can use RETURNN without rf.

> Tracing (`torch.jit.trace`) should anyway always work, except it would not cover dynamic control flow.

I understood that tracing does not even allow for dynamic time axes, but I might be wrong.

> `torch.compile` for ONNX will be the future but does not work yet?

I could find no hints that this works, but as the release of `torch.compile` was only 2 weeks ago this is not surprising. I can do some checks this or the next week, if simply passing the `torch.compile` result to `torch.onnx.export` works.

> Maybe also OpenXLA instead of ONNX.

I still do not understand how they relate to each other. For me it seems that ONNX is a standalone runtime library while XLA is a graph compiler that has to be used with a given backend/runtime.

> Any other reasonable way to get PT models running within RASR? We can also simply use the RASR-Python bridge.

While this is possible, using a RASR interface that only requires a single model file is much cleaner.

> Another RF backend just specifically for TorchScript export?

Lets just hope `torch.compile` will be able to do the same in the future.
 
> (Some earlier discussion: [#1120 (comment)](https://github.com/rwth-i6/returnn/issues/1120#issuecomment-1484230901), [#1120 (comment)](https://github.com/rwth-i6/returnn/issues/1120#issuecomment-1484253666), [#1120 (comment)](https://github.com/rwth-i6/returnn/issues/1120#issuecomment-1484714695))

**Comment by albertz** [comment]
> > Maybe also OpenXLA instead of ONNX.
> 
> I still do not understand how they relate to each other. For me it seems that ONNX is a standalone runtime library while XLA is a graph compiler that has to be used with a given backend/runtime.

Both projects provide multiple things, and very much the same:

* First of all, both define a format, which represents the computation graph, or binary. One is the XLA format, the other is the ONNX format.
* Both come with optimizers on this format, or transformations on this format.
* Both come with compilers on this format, either ahead-of-time, or on-the-fly.
* Both come with some runtime.

**Comment by patrick-wilken** [comment]
This is the runtime? https://github.com/openxla/iree Not really documented so far. 😅

**Comment by albertz** [comment]
> This is the runtime? https://github.com/openxla/iree Not really documented so far. 😅

[Here](https://openxla.github.io/iree/#project-architecture) seems to be quite some extensive documentation.

It's probably more common in case of XLA to use AOT compilation for whatever hardware you have. But this is done automatically by the XLA framework, and then executed, so from the outside, it's the same interface as a XLA runtime.

XLA is already used quite a lot, e.g. by TensorFlow (optional), by JAX (default), by PyTorch (optional, mostly for TPU).

**Comment by JackTemaki** [comment]
> [Here](https://openxla.github.io/iree/#project-architecture) seems to be quite some extensive documentation.

"IREE is in the early stages of development and is not yet ready for broad adoption"

But then the thing is adopting another interface is not too much work. ONNX support for RASR was fast, and ONNX export via TF or PyTorch worked quickly as well. But as ONNX is implemented, and other APIs are not, I do not see why we should not go for ONNX as long as it gives us all we need.

**Comment by albertz** [comment]
IREE might be in early development, but XLA is around for a long time, and very mature and stable, and used heavily by many different frameworks (TensorFlow, PyTorch, JAX). I have not really looked into the details. I would probably look at JAX, as XLA is basically the only real backend they have, so they heavily depend on it, and they use it standalone.

XLA seems overall to be a much more mature framework compared to ONNX, and now also Meta with PyTorch supports OpenXLA, so I would assume that it becomes much better supported than ONNX in the future.

My point was just that XLA is an alternative to ONNX, and probably will be better supported than ONNX in the future. We should keep that in mind and don't get too much focused on just ONNX.

**Comment by JackTemaki** [comment]
> XLA seems overall to be a much more mature framework compared to ONNX.

I have no idea how you get this impression, especially not for our purpose. XLA does not have direct runtime API and apparently uses XRT or PJRT. For all the subprojects of OpenXLA there is no documentation yet. 

E.g.: https://github.com/openxla/xla/discussions/13

"XLA on the other hand is an end-to-end "ready-to-use" ML compiler: we're working to define precisely the input graphs (with StableHLO) as well as the runtime API (likely PJRT). Behind these APIs, we're using more and more of MLIR. You may consider that the CPU/GPU flows inside XLA will be reusing all the possible components of MLIR to deliver the promise of the XLA API."

So all of this OpenXLA environment seems to be "work in progress".

Compared to that, ONNX is a very mature interchange format with a very precise and clear operation specification: https://onnx.ai/onnx/operators/index.html

And the runtime has
- A well documented API for many target languages: https://onnxruntime.ai/docs/api/
- And support for all platforms from web-assembly over mobile devices to CPU and GPU:  https://onnxruntime.ai/docs/execution-providers/

Nothing like this can be found for XLA, so this is for me an absolute no-go right now.

**Comment by albertz** [comment]
You are only looking at some of the repos of OpenXLA. As I said, XLA has a long history, and is used a lot, by TensorFlow, JAX, PyTorch. It is very mature. It is stable. It is even the main backend for JAX. In TensorFlow it's also used by default in certain cases. In PyTorch, it is also already used. Don't just look at OpenXLA. OpenXLA is just the beginning to make the format and tools more independent from Google.

**Comment by JackTemaki** [comment]
> You are only looking at some of the repos of OpenXLA. As I said, XLA has a long history, and is used a lot, by TensorFlow, JAX, PyTorch. It is very mature. It is stable. It is even the main backend for JAX. In TensorFlow it's also used by default in certain cases. In PyTorch, it is also already used. Don't just look at OpenXLA. OpenXLA is just the beginning to make the format and tools more independent from Google.

XLA in itself is a graph compiler. It is NOT a runtime. I do not doubt that the XLA compiler is not stable (while I still say documentation is.... non-existant), but I do doubt it does what we want. PJRT is the runtime that uses XLA compiled models to run it for JAX. PyTorch supports it for the sake of running on TPU runtimes. 

E.g. from PyTorch: "The PyTorch/XLA team is currently migrating from the currently-supported XRT runtime to the PjRt runtime."

So again: how is XLA useful if we want to export models and load it in "foreign" software? The full OpenXLA stack (XLA, StableHLO, MLIR, IREE) is what will give the full features of compiling the graph, transferring the model, and putting it into a runtime in another software. But again, this does not even fully exist yet and is work in progress.

**Comment by albertz** [comment]
I'm not really sure what you mean. XLA is used by many other projects. Already now. Already since a few years. It is not work-in-progress, it fully exists, and fully works. E.g. it is used by JAX. Or by TensorFlow itself. Also by Julia. Or by [Elixir in Nx / EXLA](https://github.com/elixir-nx/nx/tree/main/exla) [Nx - XLA](https://github.com/elixir-nx/xla#xla_target). Probably more. In all cases, those frameworks generate XLA code (HLO or StableHLO), and then it gets compiled and executed (all of that on-the-fly). That all fully exists now, since a few years, and works, and is used a lot.

As I understand it, OpenXLA just wants to reorganize this to become independent from TensorFlow, where you currently have the main code ([here](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/xla)). This just means that under OpenXLA, you might not find a complete solution yet, but this does not say that XLA can be (and is) already used now, and you can put a full solution together using XLA (just as in those other projects), where you export models  as XLA/HLO and then load and execute them. (Those other projects usually do that on-the-fly, they don't really store HLO on the disk and then load it in a separate process - instead, it's just one go.)

How exactly is this different to what ONNX does in the end? You have ONNX code and then it gets executed. I don't see the difference. Sure, there are technical difference, but from the outside, you have XLA (HLO) or ONNX code, and it gets executed. How is this not what we want?

Btw, maybe read here: https://news.ycombinator.com/item?id=35078791

Anyway, really my only point is that I think XLA is also very relevant and we should consider it as an alternative. We should also think about what solution will probably be more supported in the future.

**Comment by JackTemaki** [comment]
> I'm not really sure what you mean

Yes, it feels we are talking about different things. So lets leave it at that.

**Comment by JackTemaki** [comment]
> So you don't get a full graph out?

I did quick test and (onnx-)graph export with torch.compile did not work yet (well, tracing works with any model but not dynamic). Onnx separates internally between torch.nn.Module and torch.jit.ScriptModule. `torch.compile` gives a `torch._dynamo.OptimizedModule`, which is just treated as regular `torch.nn.Module` by the ONNX exporter.

**Comment by albertz** [comment]
I'm not so sure what they mean by "data-dependent control flow". I assume they do support control flow, so control flow based on tensor values. Maybe they mean by "data-dependent" that it depends on other unrelated data? In some of the links, you will see that they very explicitly say that this is going to be an alternative way to compile to TorchScript, and the old way (`torch.jit.script`/`torch.jit.trace`) would become deprecated in favor of `torch.compile`. But this is currently work-in-progress, not ready yet.

But I don't really understand too much the discussion here. We can make everything exportable if we put enough effort into it. If I'm right and `Tensor`/`Dim` are not exportable via `torch.jit.script`, we anyway need a separate RF backend just for the export, but this is doable. Or we use `torch.jit.trace`, which always works (except it doesn't handle control flow). Or if this discussion here is not about the RF at all, then it is not related to RETURNN at all. In all cases, we don't need to do anything special.

And also, as said, we can always use the RASR-Python bridge. I still don't see why this is such a big issue. If you want to have that all in a single file, this is also possible. We can just put it into a zip file. Even including the whole Python virtual env including PyTorch, if this is really what you want.

**Comment by JackTemaki** [comment]
torch.jit.trace works for the Hybrid case where no loops exist. When and how we will have Transducer setups is anyway unclear, and probably the state management / recurrency  will then be done by RASR and not via the graph. For exporting/converting there is no further code needed in RETURNN, this can be done externally, so we can close this.

A dedicated scripting backend for RF seems unnecessary, so lets just wait what PyTorch comes up with in the future and how torch.compile will be supported.

**Comment by albertz** [comment]
Btw, it was not obvious how dynamic shapes are supported via `torch.jit.trace`.

However, we noticed that `torch.onnx.export` by default internally calls `torch.jit.trace`, and `torch.onnx.export` has an option to support dynamic shapes, by specifying `dynamic_axes`.
